clusterloader2/pkg/measurement/common/pod_startup_latency.go

```diff
@@ -0,0 +1,302 @@
+		}
+	}
+	return nil
+}
+
+func checkPodFunc(p *podStartupLatencyMeasurement) func(interface{}) {
+	p.createTimes = make(map[string]map[string]metav1.Time)
+	p.scheduleTimes = make(map[string]map[string]metav1.Time)
+	p.runTimes = make(map[string]map[string]metav1.Time)
+	p.watchTimes = make(map[string]map[string]metav1.Time)
+	var mutex sync.Mutex
+	return func(obj interface{}) {
+		pod, ok := obj.(*corev1.Pod)
+		if !ok {
+			return
+		}
+		if pod.Status.Phase == corev1.PodRunning {
+			mutex.Lock()
+			defer mutex.Unlock()
+			if _, found := getNamespaceDict(p.watchTimes, pod.Namespace)[pod.Name]; !found {
+				getNamespaceDict(p.watchTimes, pod.Namespace)[pod.Name] = metav1.Now()
+				getNamespaceDict(p.createTimes, pod.Namespace)[pod.Name] = pod.CreationTimestamp
+				var startTime metav1.Time
+				for _, cs := range pod.Status.ContainerStatuses {
+					if cs.State.Running != nil {
+						if startTime.Before(&cs.State.Running.StartedAt) {
+							startTime = cs.State.Running.StartedAt
+						}
+					}
+				}
+				if startTime != metav1.NewTime(time.Time{}) {
+					getNamespaceDict(p.runTimes, pod.Namespace)[pod.Name] = startTime
+				} else {
+					glog.Errorf("Pod %v (%v) is reported to be running, but none of its containers is", pod.Name, pod.Namespace)
+				}
+			}
+		}
+	}
+}
+
+type durationSlice []time.Duration
+
+func (a durationSlice) Len() int           { return len(a) }
+func (a durationSlice) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a durationSlice) Less(i, j int) bool { return a[i] < a[j] }
+
+type podStartupLatency struct {
+	CreateToScheduleLatency measurementutil.LatencyMetric `json:"createToScheduleLatency"`
+	ScheduleToRunLatency    measurementutil.LatencyMetric `json:"scheduleToRunLatency"`
+	RunToWatchLatency       measurementutil.LatencyMetric `json:"runToWatchLatency"`
+	ScheduleToWatchLatency  measurementutil.LatencyMetric `json:"scheduleToWatchLatency"`
+	E2ELatency              measurementutil.LatencyMetric `json:"e2eLatency"`
+}
+
+// SummaryName returns name of the summary.
+func (p *podStartupLatency) SummaryName() string {
+	return podStartupLatencyMetricName
+}
+
+// PrintSummary returns summary as a string.
+func (p *podStartupLatency) PrintSummary() (string, error) {
+	return util.PrettyPrintJSON(podStartupLatencyToPerfData(p))
+}
+
+func extractLatencyMetrics(latencies []time.Duration) measurementutil.LatencyMetric {
+	length := len(latencies)
+	perc50 := latencies[int(math.Ceil(float64(length*50)/100))-1]
+	perc90 := latencies[int(math.Ceil(float64(length*90)/100))-1]
+	perc99 := latencies[int(math.Ceil(float64(length*99)/100))-1]
+	return measurementutil.LatencyMetric{Perc50: perc50, Perc90: perc90, Perc99: perc99}
+}
+
+func latencyToPerfData(l measurementutil.LatencyMetric, name string) measurementutil.DataItem {
+	return measurementutil.DataItem{
+		Data: map[string]float64{
+			"Perc50": float64(l.Perc50) / 1000000, // us -> ms
```

Comment:
us -> ms - would be 1000

it's probably ns -> ms, right?

Reply:
You are right. Fixed.